Migrator/QoI: Function input type: Don't fix Void to (Void), but fix …

bitjammer opened the pull request:
…(Void) to ()

```swift
func foo(f: Void) -> ()) {}
```

This compiler currently suggests we change this to:

```swift
func foo(f: (Void) -> ()) {}
```

That's `(()) -> ()`, almost certainly not what someone wants in Swift
4. We should suggest changing that to:
```swift
func foo(f: () -> ()) {}
```

Additionally,

```swift
func foo(f: (Void) -> ()) {}
```

Should trigger a warning that the `(Void)` input type is `(())`, and you
can't supply `()` to it in Swift 4, and suggest a fix-it change it to:

```swift
func foo(f: () -> ()) {}
```

rdar://problem/32143617

bitjammer commented:
@swift-ci Please smoke test

dabrahams commented:
IMO ()->Void is much more likely to make sense to beginners

Sent from my moss-covered three-handled family gradunza

> On May 13, 2017, at 4:52 PM, David Farler <notifications@github.com> wrote:
> 
> @swift-ci Please smoke test
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

bitjammer commented:
@swift-ci Please smoke test

bitjammer commented:
At this point, I find `Void` in general confusing and kind of wish it didn't exist, especially with SE-0110. If we never let anyone write `()`, and only `Void` (better spelled `Nothing`, IMO), I would probably agree though. Regardless, suggesting a fix-it on the return type would probably be better in a different patch.

bitjammer committed:
Migrator/QoI: Function input type: Don't fix Void to (Void), but fix (Void) to ()

```swift
func foo(f: Void) -> ()) {}
```

This compiler currently suggests we change this to:

```swift
func foo(f: (Void) -> ()) {}
```

That's `(()) -> ()`, almost certainly not what someone wants in Swift
4. We should suggest changing that to:
```swift
func foo(f: () -> ()) {}
```

Additionally,

```swift
func foo(f: (Void) -> ()) {}
```

Should trigger a warning that the `(Void)` input type is `(())`, and you
can't supply `()` to it in Swift 4, and suggest a fix-it change it to:

```swift
func foo(f: () -> ()) {}
```

rdar://problem/32143617

bitjammer commented:
@swift-ci Please smoke test

bitjammer commented:
Unrelated llbuild failure.
